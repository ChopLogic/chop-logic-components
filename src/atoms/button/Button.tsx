import { ChopLogicIcon } from '@atoms';
import { ButtonView } from '@enums';
import { withTooltip } from '@hocs';
import { ButtonProps } from '@models';
import { getClassName } from '@utils';
import { FC } from 'react';

import styles from './Button.module.scss';
import { IconButton } from './icon-button/IconButton';
import { InnerButton } from './inner-button/InnerButton';
import { PrimaryButton } from './primary-button/PrimaryButton';
import { SecondaryButton } from './secondary-button/SecondaryButton';

const ButtonSelector: FC<ButtonProps> = ({
  onClick,
  text,
  icon,
  label,
  disabled = false,
  extended = false,
  type = 'button',
  view = ButtonView.Primary,
  className,
  ...rest
}) => {
  const extendedClass = getClassName([className, { [styles.extended]: extended }]);

  switch (view) {
    case ButtonView.Primary:
      return (
        <PrimaryButton aria-label={label} type={type} onClick={onClick} disabled={disabled} className={extendedClass} {...rest}>
          <ChopLogicIcon name={icon} />
          {text}
        </PrimaryButton>
      );
    case ButtonView.Secondary:
      return (
        <SecondaryButton aria-label={label} type={type} onClick={onClick} disabled={disabled} className={extendedClass} {...rest}>
          <ChopLogicIcon name={icon} />
          {text}
        </SecondaryButton>
      );
    case ButtonView.Icon:
      return (
        <IconButton aria-label={label} type={type} onClick={onClick} disabled={disabled} {...rest} className={className}>
          <ChopLogicIcon name={icon} />
        </IconButton>
      );
    case ButtonView.Inner:
      return <InnerButton label={label} icon={icon} onClick={onClick} disabled={disabled} {...rest} className={className}></InnerButton>;
    default:
      return null;
  }
};

const Button = withTooltip(ButtonSelector);

export default Button;
