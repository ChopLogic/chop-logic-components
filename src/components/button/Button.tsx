import React, { MouseEventHandler } from 'react';
import styles from './styles.module.css';

export type ChopLogicButtonProps = {
  text?: string;
  disabled?: boolean;
  onClick?: MouseEventHandler<HTMLButtonElement>;
};

const Button: React.FC<ChopLogicButtonProps> = ({ disabled, text, onClick, ...props }) => {
  return (
    <button type='button' className={styles.button} onClick={onClick} disabled={disabled} {...props}>
      {text}
    </button>
  );
};

export default Button;
