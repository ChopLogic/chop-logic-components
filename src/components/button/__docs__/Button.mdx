import { Canvas, Meta } from '@storybook/blocks';
import Example from './Example.tsx';
import * as Button from './Button.stories.tsx';

<Meta of={Button} title='Button' />

# ChopLogicButton

A simple button component with different props.

#### Example

<Canvas of={Button.Primary} />

## Usage

```ts
import { ChopLogicButton } from "chop-logic-components";

const Example = () => {
  return (
      <ChopLogicButton
        text="Button"
        onClick={()=> console.log("Clicked")}
        type='submit'
      />
  );
};

export default Example;
```

#### Props

- **mode** - An appearance mode of the button. Can be `primary` (by default) or `secondary` (a clickable text).
- **text** _`() => void`_ - A string that represents the text content of the button.
- **disabled** - A boolean indicating whether the button should be disabled or not. When disabled, the button cannot be clicked or interacted with.
- **onClick** - A function that is called when the button is clicked. It receives a MouseEventHandler for handling the click event on the button element. It could also be a callback without arguments.
- **type** - A button type that can be `button`, `submit` or `reset` (`button` by default).
