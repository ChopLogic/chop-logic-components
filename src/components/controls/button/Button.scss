@import 'styles/global.scss';

.cl-button {
  min-height: $min-element-size;
  min-width: $min-element-size;
  border: none;
  cursor: pointer;
  padding: 0 0.75rem;
  font-size: 1rem;
  font-family: $core-font;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: $small-gap;
  overflow: hidden;

  &:focus-visible {
    outline: $focus-outline;
    outline-offset: 2px;
  }

  &__text {
    overflow: hidden;
    text-overflow: ellipsis;
  }

  // Primary styles
  &_primary {
    border-radius: $block-border-radius;
    color: $cl-color-ivory;
    box-shadow: $box-shadow;
    filter: brightness(100%);
    transition: 0.3s;
    background: $cl-primary-gradient;

    &:hover,
    &:active {
      filter: brightness(130%);
    }
  }

  // Secondary styles
  &_secondary {
    background: none;
    color: $cl-color-black;
    text-shadow: none;
    transition: 0.3s;

    & .cl-button__text {
      text-decoration: underline;
    }

    &:hover,
    &:active {
      text-shadow: $text-shadow;
    }

    &.cl-button_disabled {
      opacity: 0.6;
    }
  }

  // Danger styles
  &_danger {
    border-radius: $block-border-radius;
    color: $cl-color-ivory;
    box-shadow: $box-shadow;
    filter: brightness(100%);
    transition: 0.3s;
    background: $cl-accent-gradient;

    &:hover,
    &:active {
      filter: brightness(130%);
    }
  }

  // Icon styles
  &_icon {
    gap: 0;
    background: transparent;
    font-weight: normal;
    font-size: 1.4rem;
    border-radius: $block-border-radius;
    color: $cl-color-celtic-blue;

    &:hover,
    &:active {
      background-color: $cl-color-tint;
    }
  }

  &_disabled {
    pointer-events: none;
    filter: grayscale(80%);
    opacity: 0.9;
  }
}
