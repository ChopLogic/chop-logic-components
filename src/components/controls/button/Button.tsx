import React, { MouseEventHandler } from 'react';
import styles from './styles.module.css';
import createClassName from 'utils/create-class-name';
import { Icon } from 'enums/icon';
import 'styles';

export type ChopLogicButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  text?: string;
  onClick?: MouseEventHandler<HTMLButtonElement> | (() => void);
  view?: 'primary' | 'secondary' | 'danger' | 'icon';
  disabled?: boolean;
  icon?: Icon;
  label?: string;
};

const Button: React.FC<ChopLogicButtonProps> = ({ disabled, onClick, text, type = 'button', view = 'primary', icon, label, ...props }) => {
  const buttonClass = createClassName([
    styles.button,
    props?.className,
    {
      [styles.primary]: view === 'primary',
      [styles.secondary]: view === 'secondary',
      [styles.danger]: view === 'danger',
      [styles.icon]: view === 'icon',
      [styles.disabled]: !!disabled,
    },
  ]);

  return (
    <button aria-label={label} type={type} className={buttonClass} onClick={disabled ? undefined : onClick} disabled={disabled} {...props}>
      {icon && <span className={icon} aria-hidden='true'></span>}
      {view !== 'icon' && <span className={styles.text}>{text}</span>}
    </button>
  );
};

export default Button;
