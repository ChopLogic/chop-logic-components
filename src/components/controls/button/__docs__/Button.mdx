import { Canvas, Meta } from '@storybook/blocks';
import Example from './ButtonExample.tsx';
import * as Button from './Button.stories.tsx';

<Meta of={Button} title='Button' />

# ChopLogicButton

A simple button component with different props.

#### Example

<Canvas of={Button.Primary} />

## Usage

```ts
import { ChopLogicButton, ChopIcon } from 'chop-logic-components';
import { createCopy } from 'utils/copy';
import { Icon } from 'chop-logic-components';

const Example = () => {
  return (
    <ChopLogicButton text="Copy" onCick={createCopy} icon={Icon.Copy} />
  );
};

export default Example;
```

#### Props

- **view** - An appearance of the button:
  - `primary`: default blue button
  - `secondary`: button in the form of clickable text
  - `danger`: red button to highlight important
  - `icon`: button without text, only icon
- **text** - A string that represents the text content of the button (`Ok` by default).
- **onClick** _`() => void`_ - A function that is called when the button is clicked. It receives a MouseEventHandler for handling the click event on the button element. It could also be a callback without arguments.
- **disabled** - A boolean indicating whether the button should be disabled or not. When disabled, the button cannot be clicked or interacted with.
- **type** - A button type that can be `button`, `submit` or `reset` (`button` by default).
- **icon** - An svg icon from ChopIcon set, which could be inserted before the button text.
- **ButtonHTMLAttributes** - all standard HTML button attributes (**name**, **value**, **form**, etc.).
- **HTMLAttributes** - all global HTML attributes (**id**, **className**, **title**, **style**, etc.).
