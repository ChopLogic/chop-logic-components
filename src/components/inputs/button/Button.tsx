import React from 'react';
import { ThemeProvider } from 'styled-components';

import { ChopLogicIcon } from '@/elements';
import { ChopLogicButtonProps } from '@/types';
import { getChopLogicTheme } from '@/utils';

import { Styled3DButton, StyledFlatButton, StyledIconButton } from './Button.styled';

const ChopLogicButton: React.FC<ChopLogicButtonProps> = ({
  onClick,
  text,
  icon,
  label,
  disabled = false,
  extended = false,
  type = 'button',
  view = '3D',
  theme,
  ...rest
}) => {
  const themeValues = getChopLogicTheme(theme);

  switch (view) {
    case '3D':
      return (
        <ThemeProvider theme={themeValues}>
          <Styled3DButton $extended={extended} aria-label={label} type={type} onClick={onClick} disabled={disabled} {...rest}>
            <span className='chop-logic-button_shadow'></span>
            <span className='chop-logic-button_edge'></span>
            <span className='chop-logic-button_front'>
              <ChopLogicIcon name={icon} />
              {text}
            </span>
          </Styled3DButton>
        </ThemeProvider>
      );
    case 'flat':
      return (
        <ThemeProvider theme={themeValues}>
          <StyledFlatButton $extended={extended} aria-label={label} type={type} onClick={onClick} disabled={disabled} {...rest}>
            <ChopLogicIcon name={icon} />
            {text}
          </StyledFlatButton>
        </ThemeProvider>
      );
    case 'icon':
      return (
        <ThemeProvider theme={themeValues}>
          <StyledIconButton aria-label={label} type={type} onClick={onClick} disabled={disabled} {...rest}>
            <ChopLogicIcon name={icon} />
          </StyledIconButton>
        </ThemeProvider>
      );
    default:
      return null;
  }
};

export default ChopLogicButton;
