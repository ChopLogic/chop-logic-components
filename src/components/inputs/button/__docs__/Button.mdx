import { Canvas, Meta } from '@storybook/blocks';
import * as ButtonStories from './Button.stories';

<Meta of={ButtonStories} title='Button' />

# ChopLogicButton

Styled button component with different view modes and a customizable icon.

### Default View

<Canvas of={ButtonStories.DefaultButton} />

### Usage

```tsx
import { ChopLogicButton, ChopLogicIconName } from 'chop-logic-components';
import { createCopy } from 'utils/copy';

const Example = () => {
  return <ChopLogicButton text='Copy' onClick={createCopy} icon={ChopLogicIconName.Copy} />;
};

export default Example;
```

### Props

| Prop          | Type                                  | Description                                                                                                                                                                              | Default     | Required |
|---------------|---------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------|----------|
| **text**      | `string`                              | A string that represents the text content of the button.                                                                                                                                 | `undefined` | No       |
| **view**      | `ChopLogicButtonView`                 | An appearance of the button.                                                                                                                                                             | `undefined` | No       |
| **onClick**   | ` MouseEventHandler`                  | A function that is called when the button is clicked. It receives a MouseEventHandler for handling the click event on the button element. It could also be a callback without arguments. | `undefined` | No       |
| **disabled**  | `boolean`                             | A boolean indicating whether the button should be disabled or not. When disabled, the button cannot be clicked or interacted with.                                                       | `undefined` | No       |
| **type**      | `'button'` or `'reset'` or `'submit'` | The default behavior of the button.                                                                                                                                                      | `'button'`  | No       |
| **icon**      | `ChopLogicIconName`                   | An svg icon from ChopLogicIcon set, which could be inserted before the button text.                                                                                                      | `undefined` | No       |
| **id**        | `string`                              | The id global attribute defines an identifier (ID) which must be unique in the whole document.                                                                                           | `undefined` | No       |
| **className** | `string`                              | The class global attribute is a list of the classes of the element.                                                                                                                      | `undefined` | No       |
| **style**     | `CSSProperties`                       | The style global attribute contains CSS styling declarations to be applied to the element.                                                                                               | `undefined` | No       |
| **tabIndex**  | `number`                              | The tabindex global attribute allows developers to make HTML elements focusable.                                                                                                         | `undefined` | No       |
| **title**     | `string`                              | The title global attribute contains text representing advisory information related to the element it belongs to.                                                                         | `undefined` | No       |
| **theme**     | `'light'` or `'dark'`                 | The component's color theme.                                                                                                                                                             | `undefined` | No       |

### Types

```ts
export type ChopLogicButtonProps = Partial<CommonInputProps> & {
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  view?: ChopLogicButtonView;
  icon?: ChopLogicIconName;
  text?: string;
  extended?: boolean;
  type?: 'button' | 'reset' | 'submit';
};

export type ChopLogicButtonView = '3D' | 'flat' | 'icon';

export type CommonInputProps = CommonComponentProps & {
  label: string;
  name: string;
  disabled?: boolean;
  required?: boolean;
};

export type CommonComponentProps = {
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  tabIndex?: number;
  title?: string;
  theme?: 'dark' | 'light';
};
```
