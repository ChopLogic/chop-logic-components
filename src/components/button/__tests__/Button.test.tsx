import React from 'react';
import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import Button from '../Button';

describe('ChopLogicButton component', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Button text='Test' id='test-id' className='testClass' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have the accessible role', () => {
    render(<Button />);
    expect(screen.getByRole('button')).toBeInTheDocument();
  });
});
