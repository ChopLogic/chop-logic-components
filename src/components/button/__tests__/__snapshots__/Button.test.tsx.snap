// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChopLogicButton component > should render correctly 1`] = `
<DocumentFragment>
  <button
    class="testClass"
    id="test-id"
    type="button"
  >
    Test
  </button>
</DocumentFragment>
`;
